runtime/src/transaction_batch.rs

```diff
@@ -46,6 +46,26 @@ impl<'a, 'b> TransactionBatch<'a, 'b> {
     pub fn needs_unlock(&self) -> bool {
         self.needs_unlock
     }
+
+    /// For every error result, if the corresponding transaction is
+    /// still locked, unlock the transaction and then record the new error.
+    pub fn unlock_failures(&mut self, transaction_results: Vec<Result<()>>) {
+        assert_eq!(self.lock_results.len(), transaction_results.len());
+        // Shouldn't happen but if a batch was marked as not needing an unlock,
+        // don't unlock failures.
+        if !self.needs_unlock() {
+            return;
+        }
+
+        let txs_and_results = transaction_results
+            .iter()
+            .enumerate()
+            .filter(|(index, result)| result.is_err() && self.lock_results[*index].is_ok())
+            .map(|(index, _)| (&self.sanitized_txs[index], &self.lock_results[index]));
+
+        self.bank.unlock_accounts(txs_and_results);
+        self.lock_results = transaction_results;
```

Comment:
So the interface of this function technically allows for replacing currently errored result, and thus non locked, with an Ok result.

I can't imagine we would intentionally do this, but wdyt of adding in a inspection of each element with debug assert:
```rust
        let txs_and_results = transaction_results
            .iter()
            .enumerate()
            .inspect(|(index, result)| debug_assert!(!(result.is_ok() && self.lock_results[index].is_err()))
            .filter(|(index, result)| result.is_err() && self.lock_results[*index].is_ok())
            .map(|(index, _)| (&self.sanitized_txs[index], &self.lock_results[index]));
```

Reply:
Yeah I think we would have other problems if a failed tx was allowed to become successful in a later step of the tx pipeline but I think this is a good sanity check

Reply:
What do you think of a regular `assert` actually? I think it's worthwhile to be explicit this is a serious violation of the API and if it were to ever become possible, we should halt rather than continue processing txs

Reply:
Yeah, regular assert seems appropriate here; it's a slight added runtime cost, but probably worth it since screwing up account locks could lead to race conditions which would be tricky to trace back to this in a production setup.

Reply:
Sounds good, I just added some comments explaining why the assert is there and the risk of unsetting a lock failure